### PR TITLE
Support customized retention policy

### PR DESCRIPTION
**Pull request by SophieGuo410**
This pr support customized retention policy interface.

### CODE REVIEW

File: ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java [hunk 1]
```diff
@@ -149,10 +151,14 @@ public DatasetDao(MySqlDataAccessor dataAccessor, MySqlAccountServiceConfig mySq
             + "where (%s IS NULL or %s > now(3)) and %s = ? and %s = ? and %s = ? and %s = ? ORDER BY %s DESC LIMIT ?, 100",
         VERSION, DELETE_TS, DATASET_VERSION_TABLE, DELETE_TS, DELETE_TS, ACCOUNT_ID, CONTAINER_ID, DATASET_NAME,
         DATASET_VERSION_STATE, LAST_MODIFIED_TIME);
-    listValidDatasetVersionsSql = String.format("select %1$s from %2$s "
+    listValidDatasetVersionsByPageSql = String.format("select %1$s from %2$s "
             + "where (%3$s IS NULL or %3$s > now(3)) and %4$s = ? and %5$s = ? and %6$s = ? and %7$s = ? and %1$s >= ? "
             + "ORDER BY %1$s ASC LIMIT ?", VERSION, DATASET_VERSION_TABLE, DELETE_TS, ACCOUNT_ID, CONTAINER_ID, DATASET_NAME,
         DATASET_VERSION_STATE);
+    listValidDatasetVersionsByListSql = String.format("select %1$s, %8$s from %2$s "
```
Comment (JingQianCloud) [inline ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java:158]: %8$s is same as %3$s. Can we remove one and use %3?

File: ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java [hunk 2]
```diff
@@ -646,41 +680,72 @@ public Page<String> listAllValidDatasetVersions(short accountId, short container
...
-    } else {
-      return new ArrayList<>();
+      switch (retentionPolicy) {
+        case DEFAULT_RETENTION_POLICY:
+          datasetVersionRecordList =
+              getDatasetVersionOutOfRetentionByDefault(accountId, containerId, datasetName, retentionCount,
+                  versionSchema);
+          dataAccessor.onSuccess(Read, System.currentTimeMillis() - startTimeMs);
+          return datasetVersionRecordList;
+        case ESPRESSO_BACKUP_RETENTION_POLICY:
```
Comment (JingQianCloud) [inline ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java:708]: Can we avoid using ESPRESSO as the name of the retention policy? Instead, can we use the "meaning" of the policy itself? 
Is the policy means "keep single version"?

File: ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java (hunk 1 above)
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java]: Sure, updated.

File: ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java (hunk 2 above)
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java]: It's not just keep single version, the detail retention policy will be implemented by espresso team, so right now I'll just rename to TIMESTAMP_BASED_BACKUP_RETENTION_POLICY.

**Review by JingQianCloud — APPROVED** [review]
Approved

File: ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java [hunk 3]
```diff
@@ -646,41 +680,69 @@ public Page<String> listAllValidDatasetVersions(short accountId, short container
...
-        dataAccessor.onException(e, Read);
-        throw e;
       }
-    } else {
-      return new ArrayList<>();
+    } catch (SQLException | AccountServiceException e) {
+      dataAccessor.onException(e, Read);
+      throw e;
+    } catch (ReflectiveOperationException e) {
+      dataAccessor.onException(e, Read);
```
Comment (justinlin-linkedin) [inline ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java:721]: let's have a metric for this error.

File: ambry-utils/src/main/java/com/github/ambry/utils/Utils.java [hunk 4]
```diff
@@ -104,6 +104,7 @@ public class Utils {
    * Default retention policy for dataset.
    */
   public static final String DEFAULT_RETENTION_POLICY = "DefaultCounterBasedPolicy";
+  public static final String TIMESTAMP_BASED_BACKUP_RETENTION_POLICY = "TimestampBasedBackUpRetentionPolicy";
```
Comment (justinlin-linkedin) [inline ambry-utils/src/main/java/com/github/ambry/utils/Utils.java:107]: not used anymore.

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java (hunk 3 above)
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java]: Added.

File: ambry-utils/src/main/java/com/github/ambry/utils/Utils.java (hunk 4 above)
Comment (SophieGuo410) [reply ambry-utils/src/main/java/com/github/ambry/utils/Utils.java]: Removed.

**Pull request merged**